plugins/SegmentEditor/SegmentEditor.php

```diff
@@ -71,6 +85,92 @@ public function getKnownSegmentsToArchiveForSite(&$segments, $idSite)
+    {
+        $segmentInfo = $this->getSegmentIfIsUnprocessed();
+        if (empty($segmentInfo)) {
+            return;
+        }
+
+        list($segment, $isSegmentToPreprocess) = $segmentInfo;
+
+        // this archive has no data, the report is for a segment that gets preprocessed, and the archive for this
+        // data does not exist. this means the data will be processed later. we let the user know so they will not
+        // be confused.
+        $model = new Model();
+        $storedSegment = $model->getSegmentByDefinition($segment->getString()) ?: null;
+
+        throw new UnprocessedSegmentException($segment, $isSegmentToPreprocess, $storedSegment);
+    }
+
+    public function onLoadingError(\Exception $ex, View $dataTableView)
+    {
+        if (!($ex instanceof UnprocessedSegmentException)) {
+            Notification\Manager::cancel(self::NO_DATA_UNPROCESSED_SEGMENT_ID);
+            return;
+        }
+
+        $segment = $ex->getSegment();
+        $segmentToPreprocess = $ex->getStoredSegment();
+
+        $segmentDisplayName = !empty($segmentToPreprocess['name']) ? $segmentToPreprocess['name'] : $segment->getString();
+
+        $view = new View('@SegmentEditor/_unprocessedSegmentMessage.twig');
+        $view->isSegmentToPreprocess = $ex->isSegmentToPreprocess();
+        $view->segmentName = $segmentDisplayName;
+        $view->visitorLogLink = '#' . Url::getCurrentQueryStringWithParametersModified([
+            'category' => 'General_Visitors',
+            'subcategory' => 'Live_VisitorLog',
+        ]);
+
+        $notification = new Notification($view->render());
+        $notification->priority = Notification::PRIORITY_HIGH;
+        $notification->context = Notification::CONTEXT_INFO;
+        $notification->flags = Notification::FLAG_NO_CLEAR;
+        $notification->type = Notification::TYPE_TRANSIENT;
+        $notification->raw = true;
+
+        unset($dataTableView->error); // don't display the API error
+        $dataTableView->notifications[self::NO_DATA_UNPROCESSED_SEGMENT_ID] = $notification;
+    }
+
+    private function getSegmentIfIsUnprocessed()
+    {
+        // TODO: if in cron archiving, return null;
+        // get idSites
+        $idSite = Common::getRequestVar('idSite', false);
+        if (empty($idSite)
+            || !is_numeric($idSite)
+        ) {
+            return null;
+        }
+
+        // get segment
+        $segment = Common::getRequestVar('segment', false);
+        if (empty($segment)) {
+            return null;
+        }
+        $segment = new Segment($segment, [$idSite]);
+
+        // get period
+        $date = Common::getRequestVar('date', false);
+        $period = Common::getRequestVar('period', false);
+        $period = Period\Factory::build($period, $date);
+
+        // check if archiving is enabled. if so, the segment should have been processed.
+        $isArchivingDisabled = Rules::isArchivingDisabledFor([$idSite], $segment, $period);
```

Comment:
Not sure if it helps but there is also a `Piwik\Segment::willBeArchived()`

Reply:
I'm not quite sure if that actually works but it might do. When archiving is disabled, then it might be archived during the cron... and only need to take a while to become available. But also only if the date is recent... I think we need to take many more things into consideration eg when the date is like 2 months back, it likely won't be archived?

Reply:
Re: `willBeArchived()` that will return true if the segment is going to be processed in the next cron (for some date), but here we need to know if the segment should have been processed already, not if it's going to be processed.

> I think we need to take many more things into consideration eg when the date is like 2 months back, it likely won't be archived?

Didn't think of that... In this case we really do need some way of knowing what's next to be archived in the CronArchive run... Would it be good enough to check if the period in the request includes the current day and if the current site has had new visits? (using RawLogDao::hasSiteVisitsBetweenTimeframe() as is done in CronArchive)

Reply:
Wouldn't we want to show a message like "This report data will become available soon"? or something similar? All good though if this is not covered in the PR.

Reply:
I'm not sure I understand what you mean, there is already a message in the notification that says "Data for this segment will become available in a few hours". Is that what you're expecting?

Reply:
In case it's still not clear why I can't use willBeArchived(), is that it won't help w/ custom segments, which will also display "no data", but will never be archived in this case.

Reply:
OK I got confused with `willBeArchived` is not helpful so I thought we wouldn't show that message. 

> "Data for this segment will become available in a few hours". Is that what you're expecting?

Yes that's what I was expecting. 

> In case it's still not clear why I can't use willBeArchived(), is that it won't help w/ custom segments, which will also display "no data", but will never be archived in this case.

Still not quite sure if I'm getting it but I reckon there are those 2 cases:
* the data is currently not there but it will become available
* the data is currently not there and it won't become available.